{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "5150cebf-c3bc-498e-81c5-ab550e23a10c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for species in Human; do\n",
    "    ln -s /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/consensus_peak_calling_${species}/pseudobulk_bed_files/Enterocytes.fragments.tsv.gz \\\n",
    "          fragment_files/${species}_enterocytes.fragments.tsv.gz\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "0f5433dd-df77-495b-904c-76b30a5e0285",
   "metadata": {},
   "outputs": [],
   "source": [
    "ln -s fragment_files/Human_enterocytes.fragments.tsv.gz lifted_fragments/Human_enterocytes.hg38.fragments.tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d4be123b-9222-43c9-8e45-9c69275482ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p fragment_files\n",
    "\n",
    "for species in Gorilla Bonobo Chimpanzee Macaque Marmoset; do\n",
    "    ln -s /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/consensus_peak_calling_${species}/pseudobulk_bed_files/Enterocytes.fragments.tsv.gz \\\n",
    "          fragment_files/${species}_enterocytes.fragments.tsv.gz\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "38dd03f0-a72c-43bc-9e63-f61228c199af",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 28K\n",
      "drwxr-x--- 2 jjanssens jjanssens-group 4.0K Jun 17 15:22 .\n",
      "drwxr-x--- 4 jjanssens jjanssens-group 4.0K Jun 17 15:21 ..\n",
      "lrwxrwxrwx 1 jjanssens jjanssens-group  148 Jun 17 15:22 Bonobo_enterocytes.fragments.tsv.gz -> /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/consensus_peak_calling_Bonobo/pseudobulk_bed_files/Enterocytes.fragments.tsv.gz\n",
      "lrwxrwxrwx 1 jjanssens jjanssens-group  152 Jun 17 15:22 Chimpanzee_enterocytes.fragments.tsv.gz -> /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/consensus_peak_calling_Chimpanzee/pseudobulk_bed_files/Enterocytes.fragments.tsv.gz\n",
      "lrwxrwxrwx 1 jjanssens jjanssens-group  149 Jun 17 15:22 Gorilla_enterocytes.fragments.tsv.gz -> /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/consensus_peak_calling_Gorilla/pseudobulk_bed_files/Enterocytes.fragments.tsv.gz\n",
      "lrwxrwxrwx 1 jjanssens jjanssens-group  149 Jun 17 15:22 Macaque_enterocytes.fragments.tsv.gz -> /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/consensus_peak_calling_Macaque/pseudobulk_bed_files/Enterocytes.fragments.tsv.gz\n",
      "lrwxrwxrwx 1 jjanssens jjanssens-group  150 Jun 17 15:22 Marmoset_enterocytes.fragments.tsv.gz -> /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/consensus_peak_calling_Marmoset/pseudobulk_bed_files/Enterocytes.fragments.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "ls -lah fragment_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6cc42e76-b1b7-4a38-80a2-07b32958fdfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "mamba activate genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "aa319525-a425-408a-bf01-9b6a734683e4",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "liftOver - Move annotations from one assembly to another\n",
      "usage:\n",
      "   liftOver oldFile map.chain newFile unMapped\n",
      "oldFile and newFile are in bed format by default, but can be in GFF and\n",
      "maybe eventually others with the appropriate flags below.\n",
      "The map.chain file has the old genome as the target and the new genome\n",
      "as the query.\n",
      "\n",
      "***********************************************************************\n",
      "WARNING: liftOver was only designed to work between different\n",
      "         assemblies of the same organism. It may not do what you want\n",
      "         if you are lifting between different organisms. If there has\n",
      "         been a rearrangement in one of the species, the size of the\n",
      "         region being mapped may change dramatically after mapping.\n",
      "***********************************************************************\n",
      "\n",
      "options:\n",
      "   -minMatch=0.N Minimum ratio of bases that must remap. Default 0.95\n",
      "   -gff  File is in gff/gtf format.  Note that the gff lines are converted\n",
      "         separately.  It would be good to have a separate check after this\n",
      "         that the lines that make up a gene model still make a plausible gene\n",
      "         after liftOver\n",
      "   -genePred - File is in genePred format\n",
      "   -sample - File is in sample format\n",
      "   -bedPlus=N - File is bed N+ format (i.e. first N fields conform to bed format)\n",
      "   -positions - File is in browser \"position\" format\n",
      "   -hasBin - File has bin value (used only with -bedPlus)\n",
      "   -tab - Separate by tabs rather than space (used only with -bedPlus)\n",
      "   -pslT - File is in psl format, map target side only\n",
      "   -ends=N - Lift the first and last N bases of each record and combine the\n",
      "             result. This is useful for lifting large regions like BAC end pairs.\n",
      "   -minBlocks=0.N Minimum ratio of alignment blocks or exons that must map\n",
      "                  (default 1.00)\n",
      "   -fudgeThick    (bed 12 or 12+ only) If thickStart/thickEnd is not mapped,\n",
      "                  use the closest mapped base.  Recommended if using \n",
      "                  -minBlocks.\n",
      "   -multiple               Allow multiple output regions\n",
      "   -noSerial               In -multiple mode, do not put a serial number in the 5th BED column\n",
      "   -minChainT, -minChainQ  Minimum chain size in target/query, when mapping\n",
      "                           to multiple output regions (default 0, 0)\n",
      "   -minSizeT               deprecated synonym for -minChainT (ENCODE compat.)\n",
      "   -minSizeQ               Min matching region size in query with -multiple.\n",
      "   -chainTable             Used with -multiple, format is db.tablename,\n",
      "                               to extend chains from net (preserves dups)\n",
      "   -errorHelp              Explain error messages\n",
      "\n"
     ]
    },
    {
     "ename": "",
     "evalue": "255",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "liftOver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "9e517575-8b3c-4e88-9184-fb56671624da",
   "metadata": {},
   "outputs": [],
   "source": [
    "mamba activate genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77ccc7b8-514f-4ec8-9130-be1a43a8ee54",
   "metadata": {},
   "source": [
    "# Test single vs multicore on Bonobo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "4f33e21d-3ec9-44f9-bd54-11323317567c",
   "metadata": {},
   "outputs": [],
   "source": [
    "species=\"Bonobo\"\n",
    "input_file=\"fragment_files/${species}_enterocytes.fragments.tsv.gz\"\n",
    "chain_file=\"/cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/genomes/chains/panPan2ToHg38.over.chain.gz\"\n",
    "output_file=\"lifted_fragments/${species}_enterocytes.hg38.fragments.tsv.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "db2df89d-40b9-4185-9112-7cf1bb6f149f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üìÇ Input: fragment_files/Bonobo_enterocytes.fragments.tsv.gz\n",
      "üîó Chain: /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/genomes/chains/panPan2ToHg38.over.chain.gz\n",
      "üì§ Output: lifted_fragments/Bonobo_enterocytes.hg38.fragments.tsv.gz\n",
      "üî¢ Lines: 250000\n",
      "üì¶ Extracting top 250000 lines...\n",
      "‚úÖ Extracted 250000 lines\n",
      "üöÄ Running liftOver...\n",
      "Reading liftover chains\n",
      "Mapping coordinates\n",
      "‚úÖ LiftOver complete: 242138 fragments lifted\n",
      "üéâ Saved lifted fragments to: lifted_fragments/Bonobo_enterocytes.hg38.fragments.tsv.gz\n",
      "üß© Creating unmapped fragment file...\n",
      "üì¶ Saved unmapped fragments to: lifted_fragments/Bonobo_enterocytes.hg38.fragments.unmapped.tsv.gz\n",
      "\tCommand being timed: \"bash liftover_fragments.sh --i fragment_files/Bonobo_enterocytes.fragments.tsv.gz --c /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/genomes/chains/panPan2ToHg38.over.chain.gz --o lifted_fragments/Bonobo_enterocytes.hg38.fragments.tsv.gz --n 250000\"\n",
      "\tUser time (seconds): 68.48\n",
      "\tSystem time (seconds): 0.58\n",
      "\tPercent of CPU this job got: 100%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 1:09.02\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 409812\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 0\n",
      "\tMinor (reclaiming a frame) page faults: 125853\n",
      "\tVoluntary context switches: 1456\n",
      "\tInvoluntary context switches: 6630\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 65128\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "/usr/bin/time -v bash liftover_fragments.sh --i $input_file --c $chain_file --o $output_file --n 250000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "1ad771ca-1145-43b7-81ef-556a3e321010",
   "metadata": {},
   "outputs": [],
   "source": [
    "species=\"Bonobo\"\n",
    "input_file=\"fragment_files/${species}_enterocytes.fragments.tsv.gz\"\n",
    "chain_file=\"/cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/genomes/chains/panPan2ToHg38.over.chain.gz\"\n",
    "output_file=\"lifted_fragments/${species}_enterocytes.hg38.fragments_par.tsv.gz\"\n",
    "bash liftover_fragments_par.sh --i $input_file --c $chain_file --o $output_file --ncpu 30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "41949617-0a61-404a-a930-0755f8da8668",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üìÇ Input: fragment_files/Bonobo_enterocytes.fragments.tsv.gz\n",
      "üîó Chain: /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/genomes/chains/panPan2ToHg38.over.chain.gz\n",
      "üì§ Output: lifted_fragments/Bonobo_enterocytes.hg38.fragments_par.tsv.gz\n",
      "üî¢ Lines: 250000\n",
      "‚öôÔ∏è  Cores: 10\n",
      "üì¶ Extracting top 250000 lines...\n",
      "‚úÖ Extracted 250000 lines\n",
      "üì§ Splitting into 10 chunks...\n",
      "üöÄ Running liftOver in parallel...\n",
      "Reading liftover chains\n",
      "Reading liftover chains\n",
      "Reading liftover chains\n",
      "Reading liftover chains\n",
      "Reading liftover chains\n",
      "Reading liftover chains\n",
      "Reading liftover chains\n",
      "Reading liftover chains\n",
      "Reading liftover chains\n",
      "Reading liftover chains\n",
      "Mapping coordinates\n",
      "Mapping coordinates\n",
      "Mapping coordinates\n",
      "Mapping coordinates\n",
      "Mapping coordinates\n",
      "Mapping coordinates\n",
      "Mapping coordinates\n",
      "Mapping coordinates\n",
      "Mapping coordinates\n",
      "Mapping coordinates\n",
      "‚úÖ All liftOver jobs completed\n",
      "üéâ Saved lifted fragments to: lifted_fragments/Bonobo_enterocytes.hg38.fragments_par.tsv.gz\n",
      "üì¶ Saved unmapped fragments to: lifted_fragments/Bonobo_enterocytes.hg38.fragments_par.unmapped.tsv.gz\n",
      "\tCommand being timed: \"bash liftover_fragments_par.sh --i fragment_files/Bonobo_enterocytes.fragments.tsv.gz --c /cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/genomes/chains/panPan2ToHg38.over.chain.gz --o lifted_fragments/Bonobo_enterocytes.hg38.fragments_par.tsv.gz --ncpu 10 --n 250000\"\n",
      "\tUser time (seconds): 89.05\n",
      "\tSystem time (seconds): 3.25\n",
      "\tPercent of CPU this job got: 782%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 0:11.79\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 409980\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 49\n",
      "\tMinor (reclaiming a frame) page faults: 1036960\n",
      "\tVoluntary context switches: 6862\n",
      "\tInvoluntary context switches: 8641\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 0\n",
      "\tFile system outputs: 105328\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n"
     ]
    }
   ],
   "source": [
    "/usr/bin/time -v bash liftover_fragments_par.sh --i $input_file --c $chain_file --o $output_file --ncpu 10 --n 250000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29abbb8a-e472-4ddd-accb-dde24d44468a",
   "metadata": {},
   "source": [
    "# compare files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "38eb0061-d3e9-406a-b5f0-a8240af35628",
   "metadata": {},
   "outputs": [],
   "source": [
    "zcmp lifted_fragments/Bonobo_enterocytes.hg38.fragments_par.tsv.gz lifted_fragments/Bonobo_enterocytes.hg38.fragments_par.tsv.gz "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "f7117859-b168-417f-9533-6a20dc7c93a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "zcmp lifted_fragments/Bonobo_enterocytes.hg38.fragments_par.tsv.gz lifted_fragments/Bonobo_enterocytes.hg38.fragments.tsv.gz "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "b4efa642-a4b8-4d4d-b23c-40a064ce9377",
   "metadata": {},
   "outputs": [],
   "source": [
    "zcmp lifted_fragments/Bonobo_enterocytes.hg38.fragments_par.unmapped.tsv.gz lifted_fragments/Bonobo_enterocytes.hg38.fragments.unmapped.tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aba5674f-8dfc-4c20-94a8-dfae2c87abdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#files are all the same (no differences detected by zcmp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "434ef93a-941c-4692-8ca8-fb735a66f043",
   "metadata": {},
   "source": [
    "# Run code in parallel on all species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "65923172-6f0f-4870-9680-582b79aeaf39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üöÄ Running liftover for Macaque...\n",
      "‚úÖ Done with Macaque\n",
      "\n",
      "üöÄ Running liftover for Chimpanzee...\n",
      "‚úÖ Done with Chimpanzee\n",
      "\n",
      "üöÄ Running liftover for Bonobo...\n",
      "‚úÖ Done with Bonobo\n",
      "\n",
      "üöÄ Running liftover for Gorilla...\n",
      "‚úÖ Done with Gorilla\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# List of species and matching chain file names\n",
    "declare -A chains\n",
    "chains[\"Bonobo\"]=\"panPan2ToHg38\"\n",
    "chains[\"Chimpanzee\"]=\"panTro5ToHg38\"\n",
    "chains[\"Gorilla\"]=\"gorGor4ToHg38\"\n",
    "chains[\"Macaque\"]=\"rheMac10ToHg38\"\n",
    "\n",
    "for species in \"${!chains[@]}\"; do\n",
    "  input_file=\"fragment_files/${species}_enterocytes.fragments.tsv.gz\"\n",
    "  chain_file=\"/cluster/project/treutlein/USERS/jjans/analysis/adult_intestine/atac/genomes/chains/${chains[$species]}.over.chain.gz\"\n",
    "  output_file=\"lifted_fragments/${species}_enterocytes.hg38.fragments_par.tsv.gz\"\n",
    "\n",
    "  echo \"üöÄ Running liftover for $species...\"\n",
    "  bash liftover_fragments_par.sh --i \"$input_file\" --c \"$chain_file\" --o \"$output_file\" --ncpu 30 &> \"lifted_fragments/${species}.log\"\n",
    "  echo \"‚úÖ Done with $species\"\n",
    "  echo\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b3fe0e6-8f59-4812-87ed-572de8411e86",
   "metadata": {},
   "source": [
    "# Marmoset two step lift over"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7d46e147-82ad-4e06-a9a4-3ed116ee2aee",
   "metadata": {},
   "outputs": [],
   "source": [
    "mamba activate genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "226453e6-4c34-46ae-86d3-a859be32a5ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "export species='Marmoset'\n",
    "input_file=\"fragment_files/${species}_enterocytes.fragments.tsv.gz\"\n",
    "chain_file1=\"/cluster/project/treutlein/USERS/jjans/data/intestine/nhp_atlas/genomes/chain_files/calJac1ToCalJac4.over.chain\" \n",
    "output_file=\"lifted_fragments/${species}_enterocytes.calJac4.fragments_par.tsv.gz\"\n",
    "\n",
    "bash liftover_fragments_parchr.sh --i \"$input_file\" --c \"$chain_file1\" --o \"$output_file\" --ncpu 30 &> \"lifted_fragments/${species}_1.log\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad22e770-e276-4b8b-ab64-5908305f5a91",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file=\"lifted_fragments/${species}_enterocytes.calJac4.fragments_par.tsv.gz\"\n",
    "chain_file2=\"/cluster/project/treutlein/USERS/jjans/data/intestine/nhp_atlas/genomes/chain_files/calJac4ToHg38.over.chain\"\n",
    "output_file=\"lifted_fragments/${species}_enterocytes.hg38.fragments_par.tsv.gz\"\n",
    "\n",
    "bash liftover_fragments_parchr.sh --i \"$input_file\" --c \"$chain_file2\" --o \"$output_file\" --ncpu 30 &> \"lifted_fragments/${species}_2.log\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb28d906-4fe3-4a1d-bc92-05eb28462c36",
   "metadata": {},
   "outputs": [],
   "source": [
    "mamba activate scatac_fragment_tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b24e6981-986f-424f-bd62-f74af9cb9b36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# === Loop through species ===\n",
    "for species in \"Marmoset\"; do\n",
    "  input_frag=\"lifted_fragments/${species}_enterocytes.hg38.fragments_par.tsv.gz\"\n",
    "  chrom_file=\"/cluster/home/jjanssens/jjans/analysis/cerebellum/genomes_new/homo_sapiens/hg38.chrom.sizes\"\n",
    "  output_bw=\"bigwigs/${species}_enterocytes.hg38.bw\"\n",
    "\n",
    "  if [[ -s \"$input_frag\" && -s \"$chrom_file\" ]]; then\n",
    "    echo \"üéØ Converting $species to bigWig...\"\n",
    "    scatac_fragment_tools bigwig \\\n",
    "      -i \"$input_frag\" \\\n",
    "      -c \"$chrom_file\" \\\n",
    "      -o \"$output_bw\" \\\n",
    "      -x \n",
    "    echo \"‚úÖ Done: $output_bw\"\n",
    "  else\n",
    "    echo \"‚ö†Ô∏è  Missing input or chrom sizes for $species ‚Äî skipping\"\n",
    "  fi\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2c42991-48e0-4381-8f14-c3b529919f15",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20dd75b8-0333-4a7f-bd0d-7edabd53e7bc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0c42c78-f3dc-461f-a906-b1cc2b30df8a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dd2ad19-e87d-4afe-a42a-996e5526800f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
